crates/wasmprinter/src/component.rs

```diff
@@ -885,6 +916,307 @@ impl Printer<'_, '_> {
+                    self.end_group()?;
+                    self.end_group()?;
+                    debug_assert_eq!(state.core.func_to_type.len(), state.core.funcs as usize);
+                    state.core.funcs += 1;
+                    state.core.func_to_type.push(None);
+                }
+                CanonicalFunction::FutureCloseWritable { ty } => {
+                    self.start_group("core func ")?;
+                    self.print_name(&state.core.func_names, state.core.funcs)?;
+                    self.result.write_str(" ")?;
+                    self.start_group("canon future.close-writable ")?;
+                    self.print_idx(&state.component.type_names, ty)?;
+                    self.end_group()?;
+                    self.end_group()?;
+                    debug_assert_eq!(state.core.func_to_type.len(), state.core.funcs as usize);
+                    state.core.funcs += 1;
+                    state.core.func_to_type.push(None);
+                }
+                CanonicalFunction::ErrorContextNew { options } => {
+                    self.start_group("core func ")?;
+                    self.print_name(&state.core.func_names, state.core.funcs)?;
+                    self.result.write_str(" ")?;
+                    self.start_group("canon error-context.new")?;
+                    self.print_canonical_options(state, &options)?;
+                    self.end_group()?;
+                    self.end_group()?;
+                    debug_assert_eq!(state.core.func_to_type.len(), state.core.funcs as usize);
+                    state.core.funcs += 1;
+                    state.core.func_to_type.push(None);
+                }
+                CanonicalFunction::ErrorContextDebugMessage { options } => {
+                    self.start_group("core func ")?;
+                    self.print_name(&state.core.func_names, state.core.funcs)?;
+                    self.result.write_str(" ")?;
+                    self.start_group("canon error-context.debug-message")?;
+                    self.print_canonical_options(state, &options)?;
+                    self.end_group()?;
+                    self.end_group()?;
+                    debug_assert_eq!(state.core.func_to_type.len(), state.core.funcs as usize);
+                    state.core.funcs += 1;
+                    state.core.func_to_type.push(None);
+                }
+                CanonicalFunction::ErrorContextDrop => {
+                    self.start_group("core func ")?;
+                    self.print_name(&state.core.func_names, state.core.funcs)?;
+                    self.result.write_str(" ")?;
+                    self.start_group("canon error-context.drop")?;
+                    self.end_group()?;
+                    self.end_group()?;
+                    debug_assert_eq!(state.core.func_to_type.len(), state.core.funcs as usize);
+                    state.core.funcs += 1;
+                    state.core.func_to_type.push(None);
```

Comment:
If you'd like there's a lot of duplication between these cases around the boilerplate of printing a core func (e.g. the initial group, the debug assert, the increment, etc). That might be good to extract into a helper perhaps to have each case here focus on just the printing-the-intrinsic part